nexus/src/external_api/http_entrypoints.rs

```diff
@@ -1827,13 +1832,16 @@ async fn ip_pool_service_range_list(
 }]
 async fn ip_pool_service_range_add(
     rqctx: RequestContext<Arc<ServerContext>>,
-    range_params: TypedBody<shared::IpRange>,
+    // The plan is to change this back to shared::IpRange once IPv6 works
+    range_params: TypedBody<shared::Ipv4Range>,
```

Comment:
I think you might be right. I was somewhat biased in thinking about this by the fact that I mostly use the API through the TypeScript client. Because TS has structural types, the IPv6 and IPv4 versions of the request body look exactly the same: they're both `{ first: string, last: string }`.

https://github.com/oxidecomputer/console/blob/5d989a703359b6fa6723f59a87ba22224e62da34/libs/api/__generated__/Api.ts#L1496-L1510

There is no special constructor either; you just do `{ first: "10.0.0.1", last: "10.0.0.2" }`, etc. But in other SDKs like Rust, changing the API for this would change the constructor you use to make the request body from `IpRange::V4(Ipv4Range::new(...))` to plain `Ipv4Range::new()`, and that is real churn. On the other hand, [the presence of the `V6` variant](https://github.com/oxidecomputer/oxide.rs/blob/a4c6a2a84d8d6221c48e8984b17da72c27149082/sdk/src/generated_sdk.rs#L10748-L10751) on the `IpRange` constructor sure does make it seem like you should be able to pass in a v6 IP, and that's misleading.

I'm genuinely on the fence. It would help if you could give me a ballpark estimate for how long you think we'll have to disallow IPv6 here. If we're talking v7 and v8 and possibly beyond that, I'd lean toward changing the API now and changing it back.

Reply:
> I'm genuinely on the fence.

Yeah, I get it. The text in the HTTP request body looks the same in either case, but the Rust code (and other similarly-typed languages) looks different. I also don't want to mislead.

> It would help if you could give me a ballpark estimate for how long you think we'll have to disallow IPv6 here. If we're talking v7 and v8 and possibly beyond that, I'd lean toward changing the API now and changing it back.

I really can't say, unfortunately. The only blocker I'm currently aware of is https://github.com/oxidecomputer/omicron/issues/5090, but it wouldn't surprise me if fixing that just uncovered the next blocker. It looks like all of the [instance-NAT-related functionality](https://github.com/oxidecomputer/omicron/blob/main/nexus/db-queries/src/db/datastore/ipv4_nat_entry.rs) is currently restricted to IPv4, so that would need to be adapted or duplicated for IPv6. Maybe @internet-diglett has a better sense of the amount of work required to get IPv6 NAT support squared away?

Reply:
🤔 For sure we need to plumb it through for NAT. A cursory glance at the downstream `dendrite` side shows that we appear to have most of the machinery we need there. We'd also need to update some of the bootstore logic so we can keep the infra ips consistent with the rest of the ip related apis.

I think we also have other network related functionality that is currently v4-only like BGP (and that doesn't have the downstream machinery yet), but I don't think that is a hard blocker. It will primarily mean that if someone wants to advertise a v6 prefix via BGP, they won't be able to.

Reply:
@david-crespo circling back on this, it seems like the biggest issue currently is that we'd basically need to roll out v6 support for all of the networking related features once we do it for one feature. Off the top of my head:
* IPv6 for instances means IPv6 external networking
* IPv6 external networking means IPv6 network services (NAT, Static Routing, BGP, etc.)
* IPv6 network services means IPv6 bootstore logic

While the pathways of much of this have already been established with the IPv4 logic, it would probably still take a bit of time to roll out properly, and I think all of us are currently occupied with other work at the moment.

Reply:
That’s very helpful and it’s enough to tip me over into thinking we should make clear in the API too that it’s IPv4 only.

Reply:
Or not! I came around to Ben's view and changed it in 16f9dad5cae8c5d37bed9a3deaec3f6dbe04ea1f so that the API stays the same and we do the rejection further down. What bothered me was less the API churn and more the inconsistency _now_ between the add range endpoints and every other endpoint that deals with IP ranges.

I hoped to do the reject as low down as possible — in the datastore function — but in order to test the u128 business in IP pool utilization with IPv6 ranges from #5258, I still needed to be able to create IPv6 ranges at the datastore layer. So I moved the check one layer up, to the nexus service layer.